src/main/java/bio/terra/service/snapshot/flight/delete/DeleteSnapshotPopAndLockDatasetStep.java

```diff
@@ -0,0 +1,97 @@
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import java.util.UUID;
+
+public class DeleteSnapshotPopAndLockDatasetStep implements Step {
+
+  private SnapshotService snapshotService;
+  private DatasetService datasetService;
+  private UUID snapshotId;
+  private AuthenticatedUserRequest authenticatedUserRequest;
+  private boolean sharedLock;
+
+  public DeleteSnapshotPopAndLockDatasetStep(
+      UUID snapshotId,
+      SnapshotService snapshotService,
+      DatasetService datasetService,
+      AuthenticatedUserRequest authenticatedUserRequest,
+      boolean sharedLock) {
+    this.snapshotId = snapshotId;
+    this.snapshotService = snapshotService;
+    this.datasetService = datasetService;
+    this.authenticatedUserRequest = authenticatedUserRequest;
+    this.sharedLock = sharedLock;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) {
+    FlightMap map = context.getWorkingMap();
+    Snapshot snapshot;
+
+    // Confirm that snapshot exists and populate info about snapshot to inform optional steps
+    try {
+      snapshot = snapshotService.retrieve(snapshotId);
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_EXISTS, true);
+      boolean hasGoogleProject =
+          snapshot.getProjectResource() != null
+              && snapshot.getProjectResource().getGoogleProjectId() != null;
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_GOOGLE_PROJECT, hasGoogleProject);
+      boolean hasAzureStorageAccount = snapshot.getStorageAccountResource() != null;
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_AZURE_STORAGE_ACCOUNT, hasAzureStorageAccount);
+    } catch (SnapshotNotFoundException snapshotNotFoundException) {
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_EXISTS, false);
+      map.put(SnapshotWorkingMapKeys.DATASET_EXISTS, false);
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_GOOGLE_PROJECT, false);
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_AZURE_STORAGE_ACCOUNT, false);
+      return StepResult.getStepResultSuccess();
+    }
+
+    // Now we've confirmed the snapshot exists, let's check on the source dataset
+    UUID datasetId = snapshot.getFirstSnapshotSource().getDataset().getId();
+    map.put(DatasetWorkingMapKeys.DATASET_ID, datasetId);
+    try {
+      datasetService.lockDataset(datasetId, context.getFlightId(), sharedLock);
+      map.put(SnapshotWorkingMapKeys.DATASET_EXISTS, true);
+    } catch (DatasetNotFoundException notFoundEx) {
+      map.put(SnapshotWorkingMapKeys.DATASET_EXISTS, false);
+    } catch (RetryQueryException | DatasetLockException e) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) {
+    FlightMap map = context.getWorkingMap();
+    boolean datasetExists = map.get(SnapshotWorkingMapKeys.DATASET_EXISTS, Boolean.class);
```

Comment:
I thought we weren't supposed to use state set in the current step to undo something done in the current step. If something happens to the step after `lockDataset()` but before the working map is written, can't we end up in a state where we don't know if we're supposed to unlock or not?

On the other hand, there's no requirement that this code use `SnapshotWorkingMapKeys.DATASET_EXISTS`; it wouldn't be hard to re-derive this state in the undo method.

Reply:
I can just move all of the logic into the try statement and remove the check on SnapshotWorkingMapKeys.DATASET_EXISTS.